### ISSUE TITLE
Add a EXTERNALLY-MANAGED file to base environment

### ISSUE THREAD

**Issue by pradyunsg** [conda/conda#12245]
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

Conda should add an EXTERNALLY-MANAGED file to its environment (at least `base`) when it does not want users modifying it with `pip`.

### Why is this needed?

> Who would benefit from this feature? Why would this add value to them?

@CAM-Gerlach, at least, since he mentioned to me that pip not breaking `base` Conda environments would be nice. I expect this is more generally useful though. ;)

> What problem does this solve?

Users accidentally breaking their base environment by running a `pip install` in it.

### What should happen?

They should get an error message, describing what they should do instead of attempting a `pip install` in the base environment.

### Additional Context

https://peps.python.org/pep-0668/#specification

**Comment by jezdez** [comment]
Nice, I've been waiting for this to happen :D

**Comment by rgommers** [comment]
+1 for this - only to `base` though, installing with `pip` into development environments is important.

**Comment by jezdez** [comment]
FWIW, I've reviewed PEP 668 at the time and talked with @geofft about it and think it correctly says that conda doesn't require a change in itself. Installing pure Python packages in conda environments with pip and other installers is totally doable and preferred by many conda users to the alternative of not having access to all packages on PyPI. Anaconda defaults and conda-forge can only provide so much in terms of conda variants of PyPI packages (so far).

Adding `EXTERNALLY-MANAGED` to conda envs after all (including the dreaded "base" (read: "system") environment) will be backward-incompatible and requires careful consideration. The sad truth is that there is a malpractice in the conda community to install stuff into the base environment, which would make this a disruptive change. But especially for security minded conda users, enabling it for the "base" environment optionally could be a significant boost in predictability.

The conda CLI doesn't encourage creating environments by default and the old conda-env package hasn't even been [fully collapsed into the conda code base](https://github.com/conda/conda/issues/11633). That said, I find the blast from the past of [PEP 704](https://github.com/python/peps/pull/2964) really compelling and will be watching closely how the Python community will take it. Maybe there will be an equivalent in conda's future.

Of course as we all know, using pip with more complex packages and envs with conda packages quickly becomes a headache, and conda doesn't have a good UI to lock down envs to only allow conda packages to be installed. It's probably worth doing for new "base" environments optionally (via a miniconda option) and optionally for regular conda envs.

`EXTERNALLY-MANAGED` might be a way to make it happen for those that want it, e.g. either with an option to miniconda:

```
bash Miniconda3-latest-Linux-x86_64.sh --lock-base ...
```

or a config option to .condarc to add the `EXTERNALLY-MANAGED` file during conda initialization or shell activation:

```ini
lock_base: true
```

I'll need to do some sleuthing to find out what the plans by the conda-forge team are and equally the Anaconda Distribution people.

Any guidance from the people here would be greatly appreciated.

**Comment by pradyunsg** [comment]
FWIW, you could use PEP 704's model for rollout -- locking down the base env starting 3.12+ or so. :)

**Comment by CAM-Gerlach** [comment]
> +1 for this - only to `base` though, installing with `pip` into development environments is important.

Yeah, for sure—that was my original suggestion, in the context of PEP 704 as well, to make equally sure that `pip` works in non-`base` envs and is disabled by default in `base`. In addition to the primary purpose of preventing users from borking their `base` env and having to wipe and reinstall their whole distribution, installing into `base` tends to have a strong negative correlation with the experience and careful usage that helps maximize the chance of pip and Conda getting along.

> `EXTERNALLY-MANAGED` might be a way to make it happen for those that want it, e.g. either with an option to miniconda:
> 
> ```
> bash Miniconda3-latest-Linux-x86_64.sh --lock-base ...
> ```
> 
> or a config option to .condarc to add the `EXTERNALLY-MANAGED` file during conda initialization or shell activation:
> 
> ```ini
> lock_base: true
> ```

Yeah, I was thinking of something like that, since this is ultimately something that the conda installation administrator should be able to override for their users—though ideally enabled by default with some new Anaconda, or at least Miniconda version (and could be sooner on other distributions like Miniforge, Mambaforge, Micromamba, since they target more advanced users).  

> FWIW, you could use PEP 704's model for rollout -- locking down the base env starting 3.12+ or so. :)

AFAIK, the version of `python` in `base` can be upgraded/downgraded by the user at any time just like anything else, and has anywhere from a modest to virtually no direct impact on the user's code depending on how much they're relying on `base` vs. other environments to run it, so I don't think it would make as much sense for Conda-based distributions to key this on the Conda version in `base` as opposed to the version of Conda or the distribution itself.

**Comment by rgommers** [comment]
> It's probably worth doing for new "base" environments optionally

Optionally does not seem useful, that's not worth the effort I'd say. No one is going to opt in to this, except perhaps a handful of people who already know not to shoot themselves in the foot.

> The sad truth is that there is a malpractice in the conda community to install stuff into the base environment, which would make this a disruptive change. But especially for security minded conda users, enabling it for the "base" environment optionally could be a significant boost in predictability.

I'm not sure I understand this argument:
- This is most important to do for beginning and average users who do not (and do not want to) understand the details of how things work.
- If really needed for some reason to install into `base` with `pip` (unlikely), it can be opted into I think - from PEP 668: _"The installer should have a way for the user to override these rules, such as a command-line flag --break-system-packages. This option should not be enabled by default and should carry some connotation that its use is risky."_
- I agree that it is indeed malpractice, so disabling it by default should not be a difficult decision.

> FWIW, I've reviewed PEP 668 at the time and talked with @geofft about it and think it correctly says that conda doesn't require a change in itself

Maybe, that could go two ways. The `base` env is not not created by `conda`, but by the installer. So I guess the first change to make is in the Anaconda/Miniconda/Miniforge installers, and perhaps in [`constructor`](https://github.com/conda/constructor)? But for existing `base` envs, `conda` and `mamba` seem like the places to make a change too, if it's desired to flip the switch there (less clear, but I'd vote "yes" to that change because many users will keep the same `base` env for years, until it breaks).

**Comment by CAM-Gerlach** [comment]
> Optionally does not seem useful, that's not worth the effort I'd say. No one is going to opt in to this, except perhaps a handful of people who already know not to shoot themselves in the foot.

Agreed. Doing this by default at least for new installs would seem to be the way to go, IMO, since:

* The only users likely to realize they can enable it would be the very same group of power users who are most aware of the problems with `pip` in `base`, most careful about using it and most able to mitigate problems if they were to occur, and thus benefit from it the least (though still benefit, to be sure).
* It is the average less-experienced user who would never look for this setting that would benefit much more from it being disabled by default with an informative message/docs link (which per PEP 668 you can customize), as it is they who are much more likely to do this without knowing the consequences, and be much less well equipt to understand what happened when things inevitably start breaking.
* The opt out is, by contrast, most useful for users and larger deployments with specialized workflow needs, whose installs are likely generally managed by sysadmins that are the most attuned to the presence of a new option.
  
Theoretically, you could also add an opt-in for all environments, which might be useful for some enterprise sysadmins who don't want their users installing PyPI packages at all.

> But for existing `base` envs, `conda` and `mamba` seem like the places to make a change too, if it's desired to flip the switch there (less clear, but I'd vote "yes" to that change because many users will keep the same `base` env for years, until it breaks).

Hmm, this is a tricky one. Let's break down the likely user-facing scenarios:

<details>

<summary>User-facing scenario breakdowns</summary>

* For users who have already contaminated their `base` environment with PyPI packages:

  * If `EXTERNALLY-MANAGED` added only on new installs:

    1. User continues using `pip` as before to install, update and uninstall
    2. Eventually, things break
    3. User reinstalls anaconda/miniconda/miniforge/mambaforge, etc.
    4. User attempts to `pip install` into the new `base`
    5. User sees a helpful message explaining how it can break their whole install, likely making the connection with what just happened previously that they want to avoid having to do again, and adjusts their workflow accordingly.
    
  * If `EXTERNALLY-MANAGED` added to existing installs:
  
    1. User suddenly prevented from using `pip` to install, update and uninstall their existing packages
    2. User frustrated that they cannot manage their existing packages, so given their options are either bypass the flag or completely reinstall everything anyway, they likely just bypass the flag as the path of much less immediate resistance
    3. Eventually, things break as before and they have to reinstall their conda distro
    4. User attempts to `pip install` into the new `base`
    5. Either the opt-out persistent across installs and they can do so, or they're at least more conditioned to opt-out when they see the message again

* For users who have a clean base environment without PyPI packages:

  * If `EXTERNALLY-MANAGED` added only on new installs:
   
    1. User may eventually use `pip install` in `base` (otherwise, no effect)
    2. Go to first scenario above
    
  * If `EXTERNALLY-MANAGED` added to existing installs:
  
    1. User may eventually `pip install` into `base` (otherwise, no effect)
    2. User sees a helpful message explaining how it can break their whole install
    3. User likely keeps the flag enabled, and installs into an environment or uses `conda`—or disables it, and ends up in the same scenario as above.
    
</details>

So, TL;DR:

* Existing users who have clean `base` environments are in exactly the same position as if they had installed their conda distro with it enabled in the first place, and thus would benefit from having it enabled right away.
* Users with contaminated base envs are better off not having the flag enabled until a reinstall:
  * If only enabled on re-install, once things do eventually break and they have to reinstall, they will likely to make the connection once they see the message and be incentivized to keep it enabled and adjust their workflow accordingly in this fresh start to avoid having to do so again.
  * If enabled immediately, they will either be forced to reinstall sooner, or (more likely) just opt-out anyway, making it more likely they will stay opted-out in the future even when they do reinstall.

Therefore, the best of both worlds only setting the flag for existing `base` environments that have no PyPI packages installed, if that's practical to do.

**Comment by pradyunsg** [comment]
This isn't related to 704, but related to 668. :)

**Comment by pradyunsg** [comment]
TBH though, I don't think it's a good idea to reference concepts with PEP numbers when we have clearer names available for them. In this case, the file name unambiguously refers to the intended behaviours.

**Comment by jezdez** [comment]
I've moved this to the next planned release since I feel we've yet to come to a conclusion from the conda side how to roll this out. So there two scenarios:

- existing base environments with or without Python packages installed via pip etc
- new installations/base environments

I think for the first we'll provide the ability to validate the existing base environment for non-conda packages. If needed, users have the choice of either clearing up their non-conda packages in the base environment (purge and restart), or disabling the feature via CLI or config option.

From a user focus, I agree we shouldn't communicate this just with PEP numbers, as the intention is to protect the integrity of the base environment, we can explain it that way better.

The scenario for new installations should be rolled out with an opt-out flag similar to the conda CLI and config option that would auto-set the feature during installation.

I'm not sure if this is a constructor feature FWIW, but something to implement in miniconda. Luckily, there is a new initiative that strives to align all installers in the ecosystem to use the same code/scripts: https://github.com/conda-incubator/installer

@jaimergp you've spent a lot of time in constructor, does the second scenario ring any alarms on your end?

**Comment by ChrisBarker-NOAA** [comment]
I think this is a good idea in the short term, however:

1) You can break your base environment with conda packages as well.

2) as noted by others, there are many users that use the base environment for their regular work, so need to be able to pip install stuff -- it shouldn't be too hard to do that.

I feel the same way about PEP 704 (requiring a virtual environment) -- it's not the right way to solve this problem.

Breaking my base environment happened to me often enough that I always set up my machines to activate a "default" environment for my work that's not base. So I can fire up a terminal, do some work, conda install or pip install something, and I won't break conda itself -- every once in while, if it gets messy, I can blow it away and start again (I have a conda_requirements.txt file that I keep around to make it easy to pick up again.

So what I'd like to see, along with, (or before) locking down the base environment, is for an initial default conda install to provide two environments -- base, where conda itself lives, and a default working environment. Then we can lock down the base environment: e.g. #11342

BTW -- Python / pip should do this too -- before implementing PEP 704

Finally, this all could be opted-in by providing a conda package that installs an  EXTERNALLY-MANAGED file to whatever environment its installed into. And it could also change pip's defaults as well, via the new configuration file in `sys.prefix` 

So you could `conda install lock-down-pip` and get your optional more-locked-down environment.

Anyway, I'm playing with these idea, and will post something about it if.when I have functional idea.

**Comment by rgommers** [comment]
@ChrisBarker-NOAA imho this is not the place for a very broad discussion about the eventual fate of the `base` environment, nor about unrelated things like PEP 704. You make a few good points, but they're not this issue. For this issue:

- the scope is introducing `EXTERNALLY-MANAGED` to the `base` environment, nothing more
- this is a good idea, and @jezdez has indicated (here and in a recent conda community meeting) that this can/will be done - and a volunteer to push this forward is welcome,
- the main/only TBD is the rollout strategy,
- as a consequence, users that are only working in `base` _are_ going to have to adapt. That's part of the point really.

**Comment by ChrisBarker-NOAA** [comment]
agreed on all counts, except maybe:

> as a consequence, users that are only working in base are going to have to adapt. That's part of the point really.

It's probably best to get this in sooner than later, but ideally we'd make it easier for folks to adapt before, or soon after, pushing this through. For a good user experience, i should be easier to do "the right thing" than "the wrong thing". And simply making it harder to do the wrong thing without making it easier to do the right thing is less than ideal.

NOTE: I'm also not thrilled with the error text pip adds -- I've put in an issue about that:

https://github.com/pypa/pip/issues/11855

And the name of that flag: "--break-system-packages" is a bit heavy-handed, but that's a separate issue.

**Comment by rgommers** [comment]
> For a good user experience, i should be easier to do "the right thing" than "the wrong thing". And simply making it harder to do the wrong thing without making it easier to do the right thing is less than ideal.

The "right thing" though is to learn about using environments. Typically one per project/application you are working on. Many users have a generic `dev` environment as well, for ad-hoc tasks, but just moving users to a "use a single env for everything you do" is not something that is best practice, and it's very much debatable whether something like that should be done at all.

The whole conda workflow experience certainly needs work, and better docs with user guidance are required too. But that's a large project, and should not be mixed into this.

**Comment by jezdez** [comment]
This'll slip the May release window

**Comment by jezdez** [comment]
Removing the milestone while we're exploring how to implement this

**Comment by pradyunsg** [comment]
This is still relevant.

**Comment by jaimergp** [comment]
This is available in `conda-pypi` after https://github.com/conda-incubator/conda-pypi/pull/7, but let's keep it open until this is a default dependency in conda.

**Comment by jjhelmus** [comment]
> Finally, this all could be opted-in by providing a conda package that installs an EXTERNALLY-MANAGED file to whatever environment its installed into.

I created a recipe that implements this, conda-forge/staged-recipes#29364
